test/tools/runner/hooks/configuration.ts

```diff
@@ -81,6 +81,12 @@ const testSkipBeforeEachHook = async function () {
   if (metadata && metadata.requires && Object.keys(metadata.requires).length > 0) {
     const failedFilter = filters.find(filter => filter.filter(this.currentTest) !== true);
     if (failedFilter) {
+      const maybeSkipReason = failedFilter.filter(this.currentTest);
```

Comment:
Our `Filter` class supports returning skip reasons, but we ignored them.

Reply:
Not a big deal but the check above has already run the filter, could we preserve the result from that? I just think it could catch us by surprise in the future if a filter runs twice on the same input.

Reply:
you might think that something like this would work:
```
    const failedFilter = filters.filter(filter => filter.filter(this.currentTest) !== true)[0];
```
but it turns out the default print statement prints which filter failed by using the constructor's name (`else` block).  I think this is useful so I decided to preserve it

You might then think we could do something like this:
```
    const failedFilter = filters.find(filter => {
       const result = filter.filter(this.currentTest);
       if (result) return { result, filter };
       return null
    })[0];
```
but that seemed less nice than just filtering again.  So I chose to just run the filter again, especially because the filters are designed to be idempotent.

Reply:
a better solution is to just ensure that filters return strings, not booleans, and then we can make the change you suggested because every failing filter will return a skip reason.

Reply:
Whatever happened to a good ol' for loop? 🙂 

```js
let failedFilter: Filter;
let shouldSkip: boolean | string;
for (const filter of filters) {
  shouldSkip = filter.filter(this.currentTest);
  if (shouldSkip) {
    failedFilter = filter;
    break;
  }
}
```

Or you can leave the `find` as is and assign `shouldSkip` in the closure.

1000 ways to reach the same goal, I think it would be nice to not double call a filter, but it causes no bugs so that works too if you wanna make each one return its name.